### PR TITLE
Handle user interactions

### PR DESCRIPTION
**Pull request by elayabharath**
+ Handle analysis tabs switch, toggle comparison data, toggle filter options, change filter options
+ Refactored the filter & chart methods to computeXYZAndRender()
+ Common UI IDs for dropdowns, checkboxes (named it as utils - please let me know if may be I can change to selectors.dart)

### CODE REVIEW

File: lib/controller.dart [hunk 1]
```diff
@@ -349,14 +362,23 @@ void command(UIAction action, Data data) {
     case UIAction.changeAnalysisTab:
       var d = data as AnalysisTabChangeData;
       _selectedAnalysisTabIndex = d.tabIndex;
+      _activeFilters = {};
+      _filterValues = {};
+      _comparisonFilterValues = {};
+      view.removeFiltersWrapper();
+      view.removeAllChartWrappers();
+      _computeFilterDropdownsAndRender();
+      _computeChartBucketsAndRender();
```
Comment (marianamarasoiu) [inline lib/controller.dart]: what's the performance of clearing up the UI and reloading everything? is there any flickering? If yes, we might need to look at incrementally updating the UI...

File: lib/view.dart [hunk 2]
```diff
@@ -107,6 +114,27 @@ void clearContentTab() {
   content.children.clear();
 }
 
+void removeFiltersWrapper() {
+  if (filtersWrapper == null) {
+    logger
+        .error('Trying to remove non existant selector #${FILTERS_WRAPPER_ID}');
```
Comment (marianamarasoiu) [inline lib/view.dart]: nit:
```suggestion
        .error('Trying to remove non-existent selector #${FILTERS_WRAPPER_ID}');
```

File: lib/view.dart [hunk 3]
```diff
@@ -107,6 +114,27 @@ void clearContentTab() {
...
+  }
+
+  filtersWrapper.remove();
+}
+
+void removeAllChartWrappers() {
+  for (var wrapper in chartWrappers) {
+    if (wrapper == null) {
+      logger.error(
+          'Trying to remove non existant selector .${CHART_WRAPPER_CLASSNAME}');
```
Comment (marianamarasoiu) [inline lib/view.dart]: nit:
```suggestion
          'Trying to remove non-existent selector .${CHART_WRAPPER_CLASSNAME}');
```

File: lib/controller.dart [hunk 4]
```diff
@@ -367,27 +389,40 @@ void command(UIAction action, Data data) {
       break;
     case UIAction.toggleActiveFilter:
       var d = data as ToggleActiveFilterData;
+      var filterDropdownID = generateFilterDropdownID(d.key);
+      var comparisonDropdownID = generateComparisonFilterDropdownID(d.key);
```
Comment (marianamarasoiu) [inline lib/controller.dart]: I don't think the controller should be doing this kind of translation for the view (ie. computing the ID of the view elements). If needed, the view can keep track of the mapping key -> DOM element, or you can just pass the key to the view, which can then compute the appropriate ID?

**Review by marianamarasoiu — COMMENTED** [review]
thanks @elayabharath - a first pass and a few comments

File: lib/controller.dart (hunk 4 above)
Comment (elayabharath) [reply lib/controller.dart]: That makes sense - have changed it to make the controller just pass the filter key and the view computes to toggle on/off the drop-downs. Also removed the previously used utils for generating the IDs.

File: lib/view.dart (hunk 3 above)
Comment (elayabharath) [reply lib/view.dart]: Fixed

File: lib/view.dart (hunk 2 above)
Comment (elayabharath) [reply lib/view.dart]: Fixed

File: lib/controller.dart (hunk 1 above)
Comment (elayabharath) [reply lib/controller.dart]: At the moment I don't see flickers as such. 

But if I have to build this in..
1. The charts need to be taken out as they are highly likely to be different values for each of the tabs. Also the chartjs doesn't refresh the "on-air" image symbols without throwing away the canvas and re-rendering, say when you want to toggle on / off comparison.
2. For the filter dropdowns, I was not able to maintain the order of filter keys as I could only append them in the end after comparing the diff of filter keys across both tabs. Just realised as I was writing this I could use hidden parameter against each of the filter rows. - Will update this PR with this functionality soon.

**Comment by elayabharath** [comment]
Thanks for the first pass @marianamarasoiu 

I have addressed the following changes
1. Removed the controller passing DOM IDs to the view for the filter elements
2. Changed the filters being thrown away and re-rendering. Now I enable / disable, hide / show the required filters
3. I'm still throwing away the charts and re-rendering. 

Could you take a look at this one more time and let me know? Thank you!

**Review by marianamarasoiu — APPROVED** [review]
looks great, thanks @elayabharath !

**Comment by elayabharath** [comment]
Thank you @marianamarasoiu

**Pull request merged**